### DISCUSSION TITLE
Triton Inference Server - Ultralytics YOLOv8 Docs

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#5371] [Docs]
# Triton Inference Server - Ultralytics YOLOv8 Docs

A step-by-step guide on integrating Ultralytics YOLOv8 with Triton Inference Server for scalable and high-performance deep learning inference deployments.

https://docs.ultralytics.com/guides/triton-inference-server/

**Comment by vchaparro** [comment]
Thank you for adding the Triton integration!

I've been testing it, and everything works well with the YOLO detection model. However, I'm encountering an issue with the YOLO segmentation model. When I instantiate the segmentation model with the task set to 'segment,' it fails during inference with the following error:

```
File ~/Escritorio/tests/YOLOv8/.yoloenv/lib/python3.8/site-packages/ultralytics/models/yolo/segment/predict.py:30, in SegmentationPredictor.postprocess(self, preds, img, orig_imgs)
     28 def postprocess(self, preds, img, orig_imgs):
     29     """Applies non-max suppression and processes detections for each image in an input batch."""
---> 30     p = ops.non_max_suppression(preds[0],
     31                                 self.args.conf,
     32                                 self.args.iou,
     33                                 agnostic=self.args.agnostic_nms,
     34                                 max_det=self.args.max_det,
     35                                 nc=len(self.model.names),
     36                                 classes=self.args.classes)
     38     if not isinstance(orig_imgs, list):  # input images are a torch.Tensor, not a list
     39         orig_imgs = ops.convert_torch2numpy_batch(orig_imgs)

File ~/Escritorio/tests/YOLOv8/.yoloenv/lib/python3.8/site-packages/ultralytics/utils/ops.py:201, in non_max_suppression(prediction, conf_thres, iou_thres, classes, agnostic, multi_label, labels, max_det, nc, max_time_img, max_nms, max_wh)
    198 prediction[..., :4] = xywh2xyxy(prediction[..., :4])  # xywh to xyxy
    200 t = time.time()
--> 201 output = [torch.zeros((0, 6 + nm), device=prediction.device)] * bs
    202 for xi, x in enumerate(prediction):  # image index, image inference
    203     # Apply constraints
    204     # x[((x[:, 2:4] < min_wh) | (x[:, 2:4] > max_wh)).any(1), 4] = 0  # width-height
    205     x = x[xc[xi]]  # confidence

RuntimeError: Trying to create tensor with negative dimension -965: [0, -965]
```

Furthermore, I've noticed that it's creating the 'detect' folder instead of 'segment' when I use save=True in the predict method. This doesn't seem to be an issue with my segmentation model because I already have the same model working correctly in Triton."

It works properly when using the local exported ONNX seg. model.

**Reply by pderrenger** [reply]
@vchaparro hello! I'm glad to hear that the Triton integration is working well for you with the YOLO detection model. Regarding the issue you're facing with the YOLO segmentation model, it seems like there might be a discrepancy in the post-processing step during inference.

The error message indicates that there's an attempt to create a tensor with a negative dimension, which is not valid. This could be due to incorrect output shape handling or an unexpected response from the model. Additionally, the creation of a 'detect' folder instead of 'segment' suggests that there might be a mix-up in the task configuration.

To address this issue, please ensure that:
- The task is correctly set to 'segment' when initializing the model.
- The model is correctly loaded with the segmentation weights.
- The input data is preprocessed according to the requirements of the segmentation model.

If the problem persists, it would be helpful to review the code where the model is instantiated and the predict method is called. Make sure that the task is explicitly set to 'segment' and that the model configuration aligns with the segmentation task.

For further troubleshooting, you can refer to the [Segmentation](https://docs.ultralytics.com/tasks/segment) section of our documentation for guidance on using segmentation models with YOLOv8. If you need more detailed assistance, please open an issue on the Ultralytics YOLOv8 repo with the relevant code snippets and error logs, and our team will be happy to help you resolve the issue. 🛠️

Remember to keep your environment updated with the latest version of our repository, as we continuously work on improvements and bug fixes. 🔄

Thank you for bringing this to our attention, and we appreciate your contribution to improving YOLOv8! 🙌

**Comment by WangQvQ** [comment]
Great contribution！

**Reply by pderrenger** [reply]
@WangQvQ thank you for the kind words! 🙏 The credit goes to the entire YOLO community and the Ultralytics team for their continuous efforts to improve and support the YOLO models. If you have any questions or need further assistance with YOLOv8, feel free to reach out. Happy coding! 😊

**Comment by xx-xxx** [comment]
does triton support  batch Inference ?what inputs' format  ?the BCHW tensor ?

**Reply by pderrenger** [reply]
@xx-xxx yes, Triton Inference Server supports batch inference, which allows you to perform inference on multiple inputs simultaneously for increased throughput. The input format typically expected by Triton for image-based models like YOLOv8 is BCHW (Batch, Channels, Height, Width) tensor. When configuring your model for Triton, you'll define the input layer to accept a batch dimension, which enables batch processing. For more detailed guidance on setting up YOLOv8 with Triton, including input formatting and configuration, you can refer to our documentation. 📘🚀

**Comment by valdojoao** [comment]
Hi,
How can I replicate the line bellow in yolov5 ?
model = YOLO(f'http://localhost:8000/yolo', task='detect')

**Reply by pderrenger** [reply]
@valdojoao to replicate the functionality of connecting to a model served on a local server in YOLOv8, you would still use the `YOLO` class from the `ultralytics` package. However, please note that direct integration with a server endpoint like `http://localhost:8000/yolo` is not a standard feature provided out-of-the-box with the YOLOv8 Python package.

If you're looking to use YOLOv8 with Triton Inference Server or another serving framework, you would typically export the YOLOv8 model to an appropriate format such as ONNX using the `export` method. Then, you would set up Triton Inference Server to serve the exported model. Once Triton is serving the model, you would interact with it using Triton's client libraries, not directly through the `YOLO` class.

Here's a general outline of the steps you would take:

1. Export the YOLOv8 model to ONNX format:
   ```python
   from ultralytics import YOLO

   # Load your YOLOv8 model
   model = YOLO('yolov8n.pt')

   # Export the model to ONNX format
   model.export(format='onnx')
   ```

2. Set up Triton Inference Server to serve the exported ONNX model.

3. Use Triton's client libraries to send requests to the Triton server and receive inferences.

For more detailed instructions on how to deploy YOLOv8 with Triton Inference Server, please refer to our documentation on the Ultralytics Docs site. 📚

Keep in mind that the actual Python code to interact with Triton will be different from the line you provided for YOLOv5, as it will involve using Triton's client API.
